### PR TITLE
simplify algebra in Year offset apply methods

### PR DESCRIPTION
**Pull request by jbrockmendel**
This PR is careful to only touch YearEnd and YearBegin classes, so should not overlap with #18263 or #18278.  But like 18278, once the algebra is simplified, it looks an awful lot like 18263.

The bad news is that the process of simplifying the dynamically-defined `_increment, `_decrement, `_rollf` functions is no fun, and I'm not aware of any good way to "show your work" to walk through the in-between steps.

- [ ] closes #xxxx
- [ ] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/_libs/tslibs/offsets.pyx [hunk 1]
```diff
@@ -428,3 +428,36 @@ cpdef datetime shift_month(datetime stamp, int months, object day_opt=None):
     else:
         raise ValueError(day_opt)
     return stamp.replace(year=year, month=month, day=day)
+
+
+cpdef int _get_day_of_month(datetime other, day_opt):
```
Comment (jreback) [inline pandas/_libs/tslibs/offsets.pyx]: de-privatize

Comment (jreback) [reply pandas/_libs/tslibs/offsets.pyx]: add doc-string (examples too!)

File: pandas/_libs/tslibs/offsets.pyx [hunk 2]
```diff
@@ -428,3 +428,36 @@ cpdef datetime shift_month(datetime stamp, int months, object day_opt=None):
...
+    else:
+        raise ValueError(day_opt)
+
+
+cpdef int roll_yearday(other, n, month, day_opt='start'):
+    """
+    Possibly increment or decrement the number of periods to shift
+    based on rollforward/rollbackward conventions.
+
+    Mirrors `roll_check` in tslib.shift_months
```
Comment (jreback) [inline pandas/_libs/tslibs/offsets.pyx]: doc-string & examples

**Review by jreback — CHANGES_REQUESTED** [review]
same question, do we have sufficient tests for this?

**Comment by jbrockmendel** [comment]
> same question, do we have sufficient tests for this?

As in #18263, I think so, yes.

**Comment by jreback** [comment]
are there unit tests for these (can you point)?  in particular the raising of get_day_of_month?

File: pandas/_libs/tslibs/offsets.pyx [hunk 3]
```diff
@@ -428,3 +428,105 @@ cpdef datetime shift_month(datetime stamp, int months, object day_opt=None):
...
+    -7
+    >>>
+    >>> other = Timestamp(2072, 8, 24, 6, 17, 18)  # after June 30
+    >>> roll_yearday(other, 5, month, day_opt)
+    5
+    >>> roll_yearday(other, -7, month, day_opt)
+    -6
+
+    """
+    # Note: The other.day < ... condition will never hold when day_opt=='start'
```
Comment (jreback) [inline pandas/_libs/tslibs/offsets.pyx]: theoretically you should assert day_opt here (because if n=0 it is not checked)

**Comment by jbrockmendel** [comment]
> are there unit tests for these (can you point)? in particular the raising of get_day_of_month?

There aren't any specific tests for these.  I'll add some.

File: pandas/_libs/tslibs/offsets.pyx (hunk 3 above)
Comment (jbrockmendel) [reply pandas/_libs/tslibs/offsets.pyx]: You mean assert that it is valid?  Sure.

BTW good call on testing the raising get_day_of_month.  The exception gets ignored without an `except -1`.

**Comment by jreback** [comment]
yeah i suspect lots of things are not unit tested so taking this opportunity to put in place :)

**Comment by jbrockmendel** [comment]
Sounds good. Another BTW: we can get coverage for _libs: https://github.com/pandas-dev/pandas/issues/12634#issuecomment-338829760

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks! I love unit tests & doc-strings!

**Pull request merged**